platform/node/src/node_request.cpp

```diff
@@ -104,43 +118,34 @@ void NodeRequest::HandleCallback(const Nan::FunctionCallbackInfo<v8::Value>& inf
                     node::Buffer::Length(data)
                 );
             } else {
+                request->unref();
                 return Nan::ThrowTypeError("Response data must be a Buffer");
             }
         }
     }
 
-    if (request->ErrorMessage()) {
-        response.error = std::make_unique<mbgl::Response::Error>(
-            mbgl::Response::Error::Reason::Other,
-            request->ErrorMessage()
-        );
-    }
-
     // Send the response object to the NodeFileSource object
     callback(response);
+    request->unref();
     info.GetReturnValue().SetUndefined();
 }
 
-void NodeRequest::Execute() {
-    v8::Local<v8::Value> argv[] = { handle() };
-
-    Nan::AsyncResource res("mbgl:execute");
-    res.runInAsyncScope(Nan::To<v8::Object>(target->handle()->GetInternalField(1)).ToLocalChecked(), "request", 1, argv);
+void NodeRequest::unref() {
+  Nan::HandleScope scope;
+  delete asyncResource;
+  asyncResource = nullptr;
+  Unref();
 }
 
-NodeRequest::NodeAsyncRequest::NodeAsyncRequest(NodeRequest* request_) : request(request_) {
-    assert(request);
+NodeAsyncRequest::NodeAsyncRequest() : request(nullptr) {}
 
-    // Make sure the JS object has a pointer to this so that it can remove
-    // its pointer in the destructor
-    request->asyncRequest = this;
-}
-
-NodeRequest::NodeAsyncRequest::~NodeAsyncRequest() {
+NodeAsyncRequest::~NodeAsyncRequest() {
     if (request) {
         // Remove the callback function because the AsyncRequest was
         // canceled and we are no longer interested in the result.
-        request->callback = {};
```

Comment:
maybe a `request->cancel()` would be a cleaner interface

Reply:
Can you also add a `MBGL_VERIFY_THREAD` to this object? We need to make sure it lives on the main thread.

Reply:
The lifetime of this request object is managed by the unique_ptr that is returned here I thought - https://github.com/mapbox/mapbox-gl-native/blob/node10_support/platform/node/src/node_map.cpp#L1240